documentation/content/pony-tutorial/api.md

```diff
@@ -0,0 +1,617 @@
+
+##### Example
+
+```
+class val Event
+  // A generic event, bearing a key and a piece of data.
+  let event_data: U32
+  let event_key: Key
+
+  new val create(data: U32, key: Key) =>
+    event_data = data
+    event_key = key
+
+class MySum is State
+  var sum_of_event_data: U32
+
+  new create(sum: U32) =>
+    sum_of_event_data = sum
+
+primitive MySumAgg is Aggregation[Event, Event, MySum]
+  fun initial_accumulator(): MySum =>
+    MySum(0)
+
+  fun update(e: Event, my_sum: MySum) =>
+    my_sum.sum_of_event_data =
+      my_sum.sum_of_event_data + e.event_data
+
+  fun combine(sum1: MySum box, sum2: MySum box): MySum =>
+    MySum(sum1.sum_of_event_data + sum2.sum_of_event_data)
+
+  fun output(key: Key, sum: MySum): (None | Event) =>
+    if sum.sum_of_event_data > 0 then
+      Event(sum.sum_of_event_data, key)
+    end
+
+  fun name(): String =>
+```
+
+### Input
+
+`In` is the object that is passed to [Computations](#computation) and [StateComputations](#statecomputation). It can be any Pony type as long as it has a `val` reference capability. It and can be as simple or as complex as you would like it to be.
+
+### Key
+
+Partition keys are `string` values.
+
+### KeyExtractor
+
+A key extractor function must implement the `KeyExtractor[In: Any val]` interface and return the appropriate [Key](#key) for `input` in it's `apply(input: In): Key`.
+
+#### Example Key Extractor
+
+An example that partitions transactions based on the user assocatied with them:
+
+```
+primitive ExtractUser
+  fun apply(t: Transaction): String =>
+    t.user()
+```
+
+### Windows
+
+For an in-depth overview of windows, see [here](/core-concepts/windows). Windows are either count-based or range-based.
+
+#### Count-based
+
+`Wallaroo.count_windows(count)`: defines count-based windows that trigger an output every `count` messages.
+
+`over[ThingToCount, SumCount, RunningTotal](MySumAgg)`: specifies the aggregation used for these windows.
+
+##### Example
+
+This defines count windows with a count of 5 over a user-defined [aggregation](#aggregation) class `MySumAgg`:
+
+```
+inputs
+  .to[SumCount](Wallaroo.count_windows(5)
+    .over[ThingToCount, SumCount, RunningTotal](MySumAgg))
+  .to_sink(sink_config))
+```
+
+#### Range-based
+
+`Wallaroo.range_windows(range)`: defines range-based windows with the specified `range`.
+
+`with_slide(slide)`: optionally specifies a slide for these windows, resulting in sliding (overlapping) windows. For example, if the slide is 3 seconds, then a new window of length `range` (specified via `Wallaroo.range_windows(range)`) will be started every 3 seconds. The default slide is equal to the range (resulting in tumbling windows).
+
+`with_delay(delay)`: optionally specifies a delay on triggering a window. This is either (1) an estimation of the maximum lateness expected for messages or (2) the lateness threshold beyond which you no longer care about messages. The default delay is 0.
+
```

Comment:
We need to add another one here: `with_late_data_policy(policy)`.

Something like:
```
You can use late data policies if you want more fine-grained control over late data than just dropping late messages. See [here](/core-concepts/windows) for more details.
```